contracts/STYLE_GUIDE.md

```diff
@@ -0,0 +1,347 @@
+- When the decision is made to use a new standard, it is no longer allowed to use the old standard for new projects.
+
+### Vendor dependencies
+
+- That’s it, vendor your Solidity dependencies. Supply chain attacks are all the rage these days. There is not yet a silver bullet for best way to vendor, it depends on the size of your project and your needs. You should be as explicit as possible about where the code comes from and make sure that this is enforced in some way; e.g. reference a hash. Some options:
+  - NPM packages work for repos already in the JavaScript ecosystem. If you go this route you should lock to a hash of the repo or use a proxy registry like GitHub Packages.
+  - Copy and paste the code into a `vendor` directory. Record attribution of the code and license in the repo along with the commit or version that you pulled the code from.
+  - Foundry uses git submodules for its dependencies. We only use the `forge-std` lib through submodules, we don’t import any non-Foundry-testing code through this method.
+
+
+# Common Behaviors
+
+### Transferring Ownership
+
+- When transferring control, whether it is of a token or a role in a contract, prefer "safe ownership" transfer patterns where the recipient must accept ownership. This avoids accidentally burning the control. This is also inline with the secure pattern of [prefer pull over push](https://consensys.github.io/smart-contract-best-practices/recommendations/#favor-pull-over-push-for-external-calls).
+
+### Call with Exact Gas
+
+- `call` accepts a gas parameter, but that parameter is a ceiling on gas usage. If a transaction does not have enough gas, `call` will simply provide as much gas as it safely can. This is unintuitive and can lead to transactions failing for unexpected reasons. We have [an implementation of `callWithExactGas`](https://github.com/smartcontractkit/chainlink/blob/075f3e2caf61b8685d2dc78714f1ee39764fda17/contracts/src/v0.8/KeeperRegistry.sol#L792) to ensure the precise gas amount requested is provided.
+
+### Sending tokens
+
+- Prefer [ERC20.safeTransfer](https://docs.openzeppelin.com/contracts/2.x/api/token/erc20#SafeERC20) over ERC20.transfer
+
+### Gas golfing
+
+- Golf your code. Make it cheap, within reason.
+  - Focus on the hot path
+  - Check the [Solidity guild page](https://www.notion.so/Solidity-Guild-642cfdcfdb0a403c9a8fdb5015ff8244?pvs=21) for recordings on gas golfing
+- Most of the cost of executing Solidity is related to reading/writing storage
+- Calling other contracts will also be costly
+- Common types to safely use are
+  - uint40 for timestamps (or uint32 if you really need the space)
+  - uint96 for link, as there are only 1b link tokens
+- prefer `++i` over `i++`
+- If you’re unsure about golfing, ask in the #tech-solidity channel
+
+# Testing
+
+- Test using Foundry, do not use Hardhat
+- Aim for 90%+ *useful* coverage as a baseline, but (near) 100% is achievable in Solidity. Always 100% test the critical path.
```

Comment:
Why is 100% not achievable? Are there any specific examples that are hard/impossible to test?

Reply:
It is, in non-solidity (imo) going for exactly 100 is often not extremely productive. Here I wanted to call out that in Solidity, 100% *is* achievable.

Well, if you exclude the fact that Foundry coverage sometimes doesn't work on libraries